packages/light-client-extension-helpers/src/background/json-rpc-provider.ts

```diff
@@ -0,0 +1,205 @@
+     */
+    const daemon = yield* Effect.gen(function* () {
+      yield* Effect.addFinalizer(() => Queue.takeAll(daemonReadySignal))
+      yield* Effect.addFinalizer(() => PubSub.publish(errorPubSub, undefined))
+
+      const chain = yield* SynchronizedRef.updateAndGetEffect(
+        chainRef,
+        (oldChain) =>
+          $(
+            Effect.tryPromise(() => client.addChain(options)),
+            Effect.tap(() =>
+              $(
+                Effect.try(() => oldChain.remove()),
+                Effect.catchAll(() => Effect.void),
+              ),
+            ),
+          ),
+      )
+
+      // chain is initialized so we can signal that the daemon is ready
+      yield* Queue.offer(daemonReadySignal, undefined)
+
+      yield* $(
+        Effect.tryPromise(() => chain.nextJsonRpcResponse()),
+        Effect.tap(Effect.logDebug),
+        Effect.tap((message) => Queue.offer(messagePubSub, message)),
+      ).pipe(Effect.forever)
+    }).pipe(
+      // by scoping here, we ensure the finalizers are run before the next retry
+      Effect.scoped,
+      Effect.retry(Schedule.spaced("1 second")),
+      Effect.forkDaemon,
+      Effect.interruptible,
+    )
+
+    const provider: JsonRpcProvider = yield* Effect.sync(() =>
+      getSyncProvider(async () => {
+        await $(
+          Queue.take(daemonReadySignal),
+          Effect.tap(() => Effect.log("JSON RPC provider initialized.")),
+          runPromise,
+        )
+
+        return (onMessage, onError) => {
+          const invokeOnMessageCallback = (message: string) =>
+            $(
+              Effect.try(() => onMessage(message)),
+              Effect.catchAll(() => Effect.void),
+            )
+
+          const invokeOnErrorCallback = $(
+            Effect.try(() => onError()),
+            Effect.catchAll(() => Effect.void),
+          )
+
+          runCallback(
+            // Daemon process to propagate messages from the message queue
+            // to the onMessage callback until the first error.
+            Effect.gen(function* () {
+              const errorStream = yield* PubSub.subscribe(errorPubSub)
+              const messageStream = yield* PubSub.subscribe(messagePubSub)
+
+              yield* $(
+                Stream.fromQueue(messageStream),
+                Stream.run(Sink.forEach(invokeOnMessageCallback)),
+              ).pipe(
+                Effect.repeat({
+                  while: () => Queue.isEmpty(errorStream),
+                  schedule: Schedule.forever,
+                }),
+              )
+            }).pipe(Effect.scoped, Effect.forkDaemon),
+          )
+
+          runCallback(
+            // Daemon process that waits for the first error and then invokes
+            // the onError callback.
+            PubSub.subscribe(errorPubSub).pipe(
+              Effect.andThen((errorMessages) => Queue.take(errorMessages)),
+              Effect.andThen(() => invokeOnErrorCallback),
+              Effect.scoped,
```

Comment:
What does `scoped` do here?

Reply:
Sets the boundary for where the subscription will automatically be cleaned up / unsubscribed.

Reply:
Is there a more explicit way to do this? It's not very obvious to me :D

Otherwise, perhaps worth a comment?